### DISCUSSION TITLE
Feature: pytest-compatible mocks for supported integrations

### DISCUSSION THREAD

**Discussion by fubuloubu** [fief-dev/fief#399] [Q&A]
Would be nice to have some easy, out-of-the-box implementations that work well with how you would typically test an integration that this Fief client ships with (FastAPI, Flask, or the library behind a Python CLI). This could even completely elide Fief in the test scenario where I want to just inject fake auth data (typically, in testing I don't necessarily care about using my real auth service data, which is separate anyways). Still thinking through what would be best for this, this is my sketch (I am using it with a few FastAPI services and a SDK/CLI python package)

FastAPI testing:
```python
# `tests/conftest.py`
import pytest
from fief_client.integrations.fastapi import MockFiefAuth

from myapp import app, auth


@pytest.fixture()
def client():
    # Adapted from https://sqlmodel.tiangolo.com/tutorial/fastapi/tests/
    app.dependency_overrides[auth] = MockFiefAuth()
    yield TestClient(app)
    app.dependency_overrides.clear()
```

For CLI testing, I'm not sure I have a well-formed need yet, but it might be good to implement something useful with how Click does testing

**Comment by frankie567** [comment]
For testing with FastAPI, I would recommend to just override the dependency. `FiefUserInfo` is a simple dictionary, so you can do something like this:
 
```py
current_user = auth.current_user()

@app.get("/protected", name="protected")
async def protected(
    user: FiefUserInfo = Depends(current_user),  
):
    return HTMLResponse(
        f"<h1>You are authenticated. Your user email is {user['email']}</h1>"
    )
```

```py
mock_user = {
    "sub": "aeeb8bfa-e8f4-4724-9427-c3d5af66190e",
    "email": "anne@bretagne.duchy",
    "tenant_id": "c91ecb7f-359c-4244-8385-51ecd6c0d06b",
    "fields": {
        "first_name": "Anne",
        "last_name": "De Bretagne"
    }
}

app.dependency_overrides[current_user] = lambda: mock_user
```

**Reply by fubuloubu** [reply]
In that example, `current_user` is an object refetence? I wonder if a slight rearchitecure might make it clearer how to do that for this purpose, because `auth.current_user()` isn't possible to override

Would `Depends(auth.current_user)` also function with how the client works currently? I think it would be possible to do a dependency override on that

**Reply by fubuloubu** [reply]
Also, for reference what I did was make an override where it checks the `request.headers` for a special test-only header containing a user ID, and check that against my UserInfo cache db table for more "representative" testing

**Reply by frankie567** [reply]
Yeah, the thing is, since FastAPI `dependency_overrides` works by reference, you'll need to create a strong reference to your dependency instance to make it work. That's why I added `current_user  = auth.current_user()`.

> Also, for reference what I did was make an override where it checks the `request.headers` for a special test-only header containing a user ID, and check that against my UserInfo cache db table for more "representative" testing

Yes, that works too :)
